Islandora-Devops/isle-dc
# Add to for Secrets to be generated before generating secrets

# Purpose of this:
In a Slack chat it was brought up that users may be running default secrets and not know it and secrets are generated on every build. In fact there was several small improvements that were needed.

Currently the build process will always generate secrets even when **USE_SECRETS** is set to **true** . This adds a simple check then prompt that prevents the code from running the generate script (re)populating the */secrets/live** directory on every make up/down.

For the sake of keeping the description simple I'm going to use the terms `make local`, `make up` or `make demo` interchangeably as a reference to the make build process. It should not matter which of these to use to build ISLE. The tests will not only test the functionality but the ability to bypass/skip a function or not do anything at all. If **USE_SECRETS** is set to **false** or all the secrets exist in the secrets/live/ directory, this will not prompt the user for input.

## 3 scenarios to test for
Test USE_SECRETS set to either true or false for these 3 scenarios. But USE_SECRETS will always result in secrets/live/ being repopulated with new secrets every time a make build process is ran. 

1. Empty secrets/live/ folder
2. secrets/live/ is missing some of the expected secrets
3. Found default values in secrets/live/ secrets

Also because this is part of the build process, you will need to run `make clean` between comparing current behavior vs new and different scenarios.

### To test that this doesn't break existing setups
1. Switch to this PR's branch
2. With Isle either running or not running but previously built with secrets populated run the `make up` command.

You will see a message informing you if you're using default values, otherwise no message.

---
Run `make clean` to reset. Copy sample.env and modify the **.env** file to USE_SECRETS=**false**.

### To test when USE_SECRETS is set to **false**
1. Run `make local`
This should automatically generate custom secrets.

#### To check this worked
Check to see if the directory **/secrets/live** is populated with secrets, most of them are **NOT** set to "**password**"

Run `make clean` to reset. Copy sample.env and modify the **.env** file to USE_SECRETS=**true** to disable new secret generation as the default action.

### Testing that answering **N**o to prompt results in no changes with USE_SECRETS is set to **true**
When the script determines there's an issue, the user is prompted to either generate or copy secrets. This first check verifies nothing happens when **n** or **N** is applied to the prompts.
1. Run `make local`
2. Would you like to generate random secrets? Run a script to create secrets? [y/N] **N**
3. Would you like to copy the default secrets? Run a script to copy secrets? [y/N] **n**
4. Would you like to copy the missing secrets from /home/don/github/isle-dc/secrets/template/? [y/N] **n**
It should exit completely.

#### To check this worked
Check to see if the directory **/secrets/live** is still empty.
 
### Testing auto generate secrets with USE_SECRETS is set to **true**
1. Run `make demo`
2. Would you like to generate random secrets? Run a script to create secrets? [y/N] **y**
This should automatically generate custom secrets just like the last test.
3. Run `make up` again
There should be no message/warnings.

#### To check this worked
Check to see if the directory **/secrets/live** is populated with secrets, most of them are **NOT** set to "**password**"


### Testing copying template secrets with USE_SECRETS is set to **true**
Run `make clean` to reset. Copy sample.env and modify the **.env** file to USE_SECRETS=**true** to disable new secret generation as the default action.

1. Run `make local`
2. Would you like to generate random secrets? Run a script to create secrets? [y/N] **N**
3. Would you like to copy the default secrets? Run a script to copy secrets? [y/N] **y**
This will copy the template/ files to the live/ directory. But it's advised to make changes to these files before proceeding. The messages should be clear (if not please suggest something). This also exits not only the bash script but also the MAKEFILE. This was does intentionally. 
 4. Run 'make up' again
This will check the config and output a message at both the beginning and the end of the build. This message will inform the user which secrets are using default values with information on where to go for further information. 
![2022-02-10_21-37](https://user-images.githubusercontent.com/2738244/153529464-8a5fbeff-4232-400d-98b0-a274b69caa70.png)
5. Run `make up` again and the message should come up again last to inform the user they are using default/insecure secrets.

### Testing with missing secrets while USE_SECRETS is set to **true**
1. Delete 1 or many files in **/secrets/live**
2. Run 'make up'
3. Would you like to copy the missing secrets from /home/don/github/isle-dc/secrets/template/? [y/N] **y**
This will copy (not generate) the missing secrets from the template directory.
4. Run `make up` again and message is set back to the warning message when a default value is detected.

Note: This points to where the documentation **should** be but that doesn't exist yet.

# Sorry for the really really long description.
